vmEntryToJavaScript does not need to use the arity check entry.

MenloDorian opened the pull request:
#### d814f4a60054948f2b0b580fae1159850f6eb97d
<pre>
vmEntryToJavaScript does not need to use the arity check entry.
<a href="https://bugs.webkit.org/show_bug.cgi?id=251289">https://bugs.webkit.org/show_bug.cgi?id=251289</a>
&lt;rdar://problem/104759274&gt;

Reviewed by Geoffrey Garen and Yusuke Suzuki.

This is because we can tell the ProtoCallFrame to ensure that there is enough argument
space to meet arity requirements instead.

With this, we can also make the following changes:

1. We no longer need to look up the arity check entry in JITCode.  Instead, we&apos;ll add
   a m_addressForCall CodePtr in the base JITCode which is common across all JITCode
   subclasses, and therefore can be accessed without a virtual call.  To enable this,
   we also change JITCodeWithCodeRef to not use a CodeRef.  Instead, we use the
   m_addressForCall CodePtr along with a m_executableMemory RefPtr in JITCodeWithCodeRef
   itself to track the equivalent information.

2. Remove JITCode::execute().  Clients will not call vmEntryToJavaScript directly instead.
   Other than calling vmEntryToJavaScript, the other things that JITCode::execute() does are:

   a. Set vm.didEnterVM on exit.  The only client that doesn&apos;t already do this is
      Interpreter::executeCachedCall.  So, we&apos;ll just add it there.

   b. Change the return value of vmEntryToJavaScript to jsNull if an exception is present
      on exit.  I believe this was in to ensure that checkedReturn() does not complain.
      The purpose of checkeReturn() is for a legacy reason: back then, we didn&apos;t have exception
      check validation to ensure that exceptions are checked in all the right places.
      checkedReturn() adds an ASSERT to help ensure that we don&apos;t return a nullptr.  However, a
      null check doesn&apos;t really provide much of a guarantee that the return value is valid.  So,
      we&apos;ll just remove checkedReturn() especially since we now have exception check validation.

3. We also change the return paths in vmEntrytoJavaScript for errors and unhandled exceptions
   to always return jsUndefined.  This is not strictly needed but it is extremely low cost, and
   helps make debugging an unchecked exception easier (as opposed to returning whatever random
   value was in the register).

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/ftl/FTLJITCode.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeCall):
(JSC::Interpreter::executeConstruct):
(JSC::Interpreter::executeEval):
(JSC::Interpreter::executeModuleProgram):
* Source/JavaScriptCore/interpreter/Interpreter.h:
(JSC::Interpreter::checkedReturn): Deleted.
* Source/JavaScriptCore/interpreter/InterpreterInlines.h:
(JSC::Interpreter::executeCachedCall):
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
(JSC::ProtoCallFrame::init):
* Source/JavaScriptCore/jit/JITCode.cpp:
(JSC::JITCode::JITCode):
(JSC::JITCodeWithCodeRef::JITCodeWithCodeRef):
(JSC::JITCodeWithCodeRef::~JITCodeWithCodeRef):
(JSC::JITCodeWithCodeRef::executableAddressAtOffset):
(JSC::JITCodeWithCodeRef::dataAddressAtOffset):
(JSC::JITCodeWithCodeRef::offsetOf):
(JSC::JITCodeWithCodeRef::size):
(JSC::JITCodeWithCodeRef::contains):
(JSC::JITCodeWithCodeRef::swapCodeRefForDebugger):
(JSC::DirectJITCode::DirectJITCode):
(JSC::DirectJITCode::initializeCodeRefForDFG):
(JSC::DirectJITCode::addressForCall):
(JSC::NativeJITCode::addressForCall):
* Source/JavaScriptCore/jit/JITCode.h:
(JSC::JITCode::addressForCall const):
* Source/JavaScriptCore/jit/JITCodeInlines.h: Removed.
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::evaluate):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:

Canonical link: <a href="https://commits.webkit.org/259520@main">https://commits.webkit.org/259520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a640cf647464df6ec54ca0517a5ea8874520088

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105073 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37963 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114336 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174523 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15291 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5074 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97394 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11822 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39328 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26449 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94918 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7490 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27808 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92958 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5220 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7584 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4394 "Found 1 new test failure: fast/images/avif-image-document.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30289 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13639 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47361 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101654 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9373 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25361 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3503 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
```diff
@@ -37,13 +37,10 @@ inline void ProtoCallFrame::init(CodeBlock* codeBlock, JSGlobalObject* globalObj
-        this->hasArityMismatch = false;
-
-    // Round up argCountIncludingThis to keep the stack frame size aligned.
-    size_t paddedArgsCount = roundArgumentCountToAlignFrame(argCountIncludingThis);
+        paddedArgsCount = codeBlock->numParameters();
+    paddedArgsCount = roundArgumentCountToAlignFrame(paddedArgsCount);
```
What is the case where CodeBlock is null? In that case, why is it OK to assume that arity matches?

On the surface, this appears to be a correctness regression. The previous code calculated 'hasArityMismatch = false', but then dropped that value on the floor -- two bugs cancelling each other out. In this patch, when codeBlock is null, we jump directly to the ArityCheckNotRequired entry point -- potentially causing the callee to dereference uninitialized argument memory.

Constellation replied:
Wasm calls, and host function calls will get CodeBlock == nullptr. And they can accept arity-mismatch cases. So this is fine. Arity-mismatch case matters only for JS calls, and they have CodeBlock.

MenloDorian replied:
When CodeBlock is null, we're looking at a Wasm or Native function.  In those cases, their normal and arity check entries both point to the same entry anyway.  They do not have an arity check version.  The wasm and native function also does not access arguments beyond the count is set in `CallFrame::argumentIncludingThis`.  This is already the case in present code.

`hasArityMismatch` is unused code.  No one checks it anymore.

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.h:
```diff
@@ -157,18 +157,6 @@ using JSOrWasmInstruction = std::variant<const JSInstruction*, const WasmInstruc
-            ASSERT(returnValue);
-            return returnValue;
-        }
-
```
Yeah, they were silly ones :P

geoffreygaren reviewed (approved):
r=me

Constellation reviewed (approved):
r=me

Commit:
vmEntryToJavaScript does not need to use the arity check entry.
https://bugs.webkit.org/show_bug.cgi?id=251289
<rdar://problem/104759274>

Reviewed by Geoffrey Garen and Yusuke Suzuki.

This is because we can tell the ProtoCallFrame to ensure that there is enough argument
space to meet arity requirements instead.

With this, we can also make the following changes:

1. We no longer need to look up the arity check entry in JITCode.  Instead, we'll add
   a m_addressForCall CodePtr in the base JITCode which is common across all JITCode
   subclasses, and therefore can be accessed without a virtual call.  To enable this,
   we also change JITCodeWithCodeRef to not use a CodeRef.  Instead, we use the
   m_addressForCall CodePtr along with a m_executableMemory RefPtr in JITCodeWithCodeRef
   itself to track the equivalent information.

2. Remove JITCode::execute().  Clients will not call vmEntryToJavaScript directly instead.
   Other than calling vmEntryToJavaScript, the other things that JITCode::execute() does are:

   a. Set vm.didEnterVM on exit.  The only client that doesn't already do this is
      Interpreter::executeCachedCall.  So, we'll just add it there.

   b. Change the return value of vmEntryToJavaScript to jsNull if an exception is present
      on exit.  I believe this was in to ensure that checkedReturn() does not complain.
      The purpose of checkeReturn() is for a legacy reason: back then, we didn't have exception
      check validation to ensure that exceptions are checked in all the right places.
      checkedReturn() adds an ASSERT to help ensure that we don't return a nullptr.  However, a
      null check doesn't really provide much of a guarantee that the return value is valid.  So,
      we'll just remove checkedReturn() especially since we now have exception check validation.

3. We also change the return paths in vmEntrytoJavaScript for errors and unhandled exceptions
   to always return jsUndefined.  This is not strictly needed but it is extremely low cost, and
   helps make debugging an unchecked exception easier (as opposed to returning whatever random
   value was in the register).

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/ftl/FTLJITCode.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeCall):
(JSC::Interpreter::executeConstruct):
(JSC::Interpreter::executeEval):
(JSC::Interpreter::executeModuleProgram):
* Source/JavaScriptCore/interpreter/Interpreter.h:
(JSC::Interpreter::checkedReturn): Deleted.
* Source/JavaScriptCore/interpreter/InterpreterInlines.h:
(JSC::Interpreter::executeCachedCall):
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
(JSC::ProtoCallFrame::init):
* Source/JavaScriptCore/jit/JITCode.cpp:
(JSC::JITCode::JITCode):
(JSC::JITCodeWithCodeRef::JITCodeWithCodeRef):
(JSC::JITCodeWithCodeRef::~JITCodeWithCodeRef):
(JSC::JITCodeWithCodeRef::executableAddressAtOffset):
(JSC::JITCodeWithCodeRef::dataAddressAtOffset):
(JSC::JITCodeWithCodeRef::offsetOf):
(JSC::JITCodeWithCodeRef::size):
(JSC::JITCodeWithCodeRef::contains):
(JSC::JITCodeWithCodeRef::swapCodeRefForDebugger):
(JSC::DirectJITCode::DirectJITCode):
(JSC::DirectJITCode::initializeCodeRefForDFG):
(JSC::DirectJITCode::addressForCall):
(JSC::NativeJITCode::addressForCall):
* Source/JavaScriptCore/jit/JITCode.h:
(JSC::JITCode::addressForCall const):
* Source/JavaScriptCore/jit/JITCodeInlines.h: Removed.
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::evaluate):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:

Canonical link: https://commits.webkit.org/259520@main
